File: fontes/backend/delphi/dmvc/Model/Vendas/VendaOrcamentoCabecalho.pas
{*******************************************************************************
Title: T2Ti ERP Fenix                                                                
Description: Model relacionado à tabela [VENDA_ORCAMENTO_CABECALHO] 
                                                                                
The MIT License                                                                 
                                                                                
Copyright: Copyright (C) 2020 T2Ti.COM                                          
                                                                                
Permission is hereby granted, free of charge, to any person                     
obtaining a copy of this software and associated documentation                  
files (the "Software"), to deal in the Software without                         
restriction, including without limitation the rights to use,                    
copy, modify, merge, publish, distribute, sublicense, and/or sell               
copies of the Software, and to permit persons to whom the                       
Software is furnished to do so, subject to the following                        
conditions:                                                                     
                                                                                
The above copyright notice and this permission notice shall be                  
included in all copies or substantial portions of the Software.                 
                                                                                
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,                 
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES                 
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND                        
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT                     
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,                    
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING                    
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR                   
OTHER DEALINGS IN THE SOFTWARE.                                                 
                                                                                
       The author may be contacted at:                                          
           t2ti.com@gmail.com                                                   
                                                                                
@author Albert Eije (alberteije@gmail.com)                    
@version 1.0.0
*******************************************************************************}
unit VendaOrcamentoCabecalho;

interface

uses
  Generics.Collections, System.SysUtils,
  VendaOrcamentoDetalhe, VendaCondicoesPagamento, Vendedor, Cliente, Transportadora, 
  MVCFramework.Serializer.Commons, ModelBase;

type

  [MVCNameCase(ncCamelCase)]
  TVendaOrcamentoCabecalho = class(TModelBase)
  private
    FId: Integer;
    FIdVendaCondicoesPagamento: Integer;
    FIdVendedor: Integer;
    FIdCliente: Integer;
    FIdTransportadora: Integer;
    FCodigo: string;
    FDataCadastro: TDateTime;
    FDataEntrega: TDateTime;
    FValidade: TDateTime;
    FTipoFrete: string;
    FValorSubtotal: Extended;
    FValorFrete: Extended;
    FTaxaComissao: Extended;
    FValorComissao: Extended;
    FTaxaDesconto: Extended;
    FValorDesconto: Extended;
    FValorTotal: Extended;
    FObservacao: string;
      
    FVendaCondicoesPagamento: TVendaCondicoesPagamento;
    FVendedor: TVendedor;
    FCliente: TCliente;
    FTransportadora: TTransportadora;
    FListaVendaOrcamentoDetalhe: TObjectList<TVendaOrcamentoDetalhe>;
  public
    constructor Create; virtual;
    destructor Destroy; override;

    [MVCColumnAttribute('ID', True)]
		[MVCNameAsAttribute('id')]
		property Id: Integer read FId write FId;
    [MVCColumnAttribute('ID_VENDA_CONDICOES_PAGAMENTO')]
		[MVCNameAsAttribute('idVendaCondicoesPagamento')]
		property IdVendaCondicoesPagamento: Integer read FIdVendaCondicoesPagamento write FIdVendaCondicoesPagamento;
    [MVCColumnAttribute('ID_VENDEDOR')]
		[MVCNameAsAttribute('idVendedor')]
		property IdVendedor: Integer read FIdVendedor write FIdVendedor;
    [MVCColumnAttribute('ID_CLIENTE')]
		[MVCNameAsAttribute('idCliente')]
		property IdCliente: Integer read FIdCliente write FIdCliente;
    [MVCColumnAttribute('ID_TRANSPORTADORA')]
		[MVCNameAsAttribute('idTransportadora')]
		property IdTransportadora: Integer read FIdTransportadora write FIdTransportadora;
    [MVCColumnAttribute('CODIGO')]
		[MVCNameAsAttribute('codigo')]
		property Codigo: string read FCodigo write FCodigo;
    [MVCColumnAttribute('DATA_CADASTRO')]
		[MVCNameAsAttribute('dataCadastro')]
		property DataCadastro: TDateTime read FDataCadastro write FDataCadastro;
    [MVCColumnAttribute('DATA_ENTREGA')]
		[MVCNameAsAttribute('dataEntrega')]
		property DataEntrega: TDateTime read FDataEntrega write FDataEntrega;
    [MVCColumnAttribute('VALIDADE')]
		[MVCNameAsAttribute('validade')]
		property Validade: TDateTime read FValidade write FValidade;
    [MVCColumnAttribute('TIPO_FRETE')]
		[MVCNameAsAttribute('tipoFrete')]
		property TipoFrete: string read FTipoFrete write FTipoFrete;
    [MVCColumnAttribute('VALOR_SUBTOTAL')]
		[MVCNameAsAttribute('valorSubtotal')]
		property ValorSubtotal: Extended read FValorSubtotal write FValorSubtotal;
    [MVCColumnAttribute('VALOR_FRETE')]
		[MVCNameAsAttribute('valorFrete')]
		property ValorFrete: Extended read FValorFrete write FValorFrete;
    [MVCColumnAttribute('TAXA_COMISSAO')]
		[MVCNameAsAttribute('taxaComissao')]
		property TaxaComissao: Extended read FTaxaComissao write FTaxaComissao;
    [MVCColumnAttribute('VALOR_COMISSAO')]
		[MVCNameAsAttribute('valorComissao')]
		property ValorComissao: Extended read FValorComissao write FValorComissao;
    [MVCColumnAttribute('TAXA_DESCONTO')]
		[MVCNameAsAttribute('taxaDesconto')]
		property TaxaDesconto: Extended read FTaxaDesconto write FTaxaDesconto;
    [MVCColumnAttribute('VALOR_DESCONTO')]
		[MVCNameAsAttribute('valorDesconto')]
		property ValorDesconto: Extended read FValorDesconto write FValorDesconto;
    [MVCColumnAttribute('VALOR_TOTAL')]
		[MVCNameAsAttribute('valorTotal')]
		property ValorTotal: Extended read FValorTotal write FValorTotal;
    [MVCColumnAttribute('OBSERVACAO')]
		[MVCNameAsAttribute('observacao')]
		property Observacao: string read FObservacao write FObservacao;
      
    [MVCNameAsAttribute('vendaCondicoesPagamento')]
		property VendaCondicoesPagamento: TVendaCondicoesPagamento read FVendaCondicoesPagamento write FVendaCondicoesPagamento;
    [MVCNameAsAttribute('vendedor')]
		property Vendedor: TVendedor read FVendedor write FVendedor;
    [MVCNameAsAttribute('cliente')]
		property Cliente: TCliente read FCliente write FCliente;
    [MVCNameAsAttribute('transportadora')]
		property Transportadora: TTransportadora read FTransportadora write FTransportadora;
    [MapperListOf(TVendaOrcamentoDetalhe)]
	[MVCNameAsAttribute('listaVendaOrcamentoDetalhe')]
	property ListaVendaOrcamentoDetalhe: TObjectList<TVendaOrcamentoDetalhe> read FListaVendaOrcamentoDetalhe write FListaVendaOrcamentoDetalhe;
  end;

implementation

{ TVendaOrcamentoCabecalho }

constructor TVendaOrcamentoCabecalho.Create;
begin
  inherited;

  FListaVendaOrcamentoDetalhe := TObjectList<TVendaOrcamentoDetalhe>.Create;
  FVendaCondicoesPagamento := TVendaCondicoesPagamento.Create;
  FVendedor := TVendedor.Create;
  FCliente := TCliente.Create;
  FTransportadora := TTransportadora.Create;
end;

destructor TVendaOrcamentoCabecalho.Destroy;
begin
  FreeAndNil(FListaVendaOrcamentoDetalhe);
  FreeAndNil(FVendaCondicoesPagamento);
  FreeAndNil(FVendedor);
  FreeAndNil(FCliente);
  FreeAndNil(FTransportadora);
  inherited;
end;



end.

File: fontes/frontend/fenix_provider/lib/src/model/vendas/venda_orcamento_cabecalho.dart
/*
Title: T2Ti ERP 3.0                                                                
Description: Model relacionado à tabela [VENDA_ORCAMENTO_CABECALHO] 
                                                                                
The MIT License                                                                 
                                                                                
Copyright: Copyright (C) 2021 T2Ti.COM                                          
                                                                                
Permission is hereby granted, free of charge, to any person                     
obtaining a copy of this software and associated documentation                  
files (the "Software"), to deal in the Software without                         
restriction, including without limitation the rights to use,                    
copy, modify, merge, publish, distribute, sublicense, and/or sell               
copies of the Software, and to permit persons to whom the                       
Software is furnished to do so, subject to the following                        
conditions:                                                                     
                                                                                
The above copyright notice and this permission notice shall be                  
included in all copies or substantial portions of the Software.                 
                                                                                
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,                 
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES                 
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND                        
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT                     
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,                    
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING                    
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR                   
OTHER DEALINGS IN THE SOFTWARE.                                                 
                                                                                
       The author may be contacted at:                                          
           t2ti.com@gmail.com                                                   
                                                                                
@author Albert Eije (alberteije@gmail.com)                    
@version 1.0.0
*******************************************************************************/
import 'dart:convert';

import 'package:intl/intl.dart';
import 'package:fenix/src/model/model.dart';

class VendaOrcamentoCabecalho {
	int? id;
	int? idVendaCondicoesPagamento;
	int? idVendedor;
	int? idCliente;
	int? idTransportadora;
	String? codigo;
	DateTime? dataCadastro;
	DateTime? dataEntrega;
	DateTime? validade;
	String? tipoFrete;
	double? valorSubtotal;
	double? valorFrete;
	double? taxaComissao;
	double? valorComissao;
	double? taxaDesconto;
	double? valorDesconto;
	double? valorTotal;
	String? observacao;
	VendaCondicoesPagamento? vendaCondicoesPagamento;
	Vendedor? vendedor;
	Cliente? cliente;
	Transportadora? transportadora;
	List<VendaOrcamentoDetalhe>? listaVendaOrcamentoDetalhe = [];

	VendaOrcamentoCabecalho({
			id,
			idVendaCondicoesPagamento,
			idVendedor,
			idCliente,
			idTransportadora,
			codigo,
			dataCadastro,
			dataEntrega,
			validade,
			tipoFrete,
			valorSubtotal = 0.0,
			valorFrete = 0.0,
			taxaComissao = 0.0,
			valorComissao = 0.0,
			taxaDesconto = 0.0,
			valorDesconto = 0.0,
			valorTotal = 0.0,
			observacao,
			vendaCondicoesPagamento,
			vendedor,
			cliente,
			transportadora,
			listaVendaOrcamentoDetalhe,
		});

	static List<String> campos = <String>[
		'ID', 
		'CODIGO', 
		'DATA_CADASTRO', 
		'DATA_ENTREGA', 
		'VALIDADE', 
		'TIPO_FRETE', 
		'VALOR_SUBTOTAL', 
		'VALOR_FRETE', 
		'TAXA_COMISSAO', 
		'VALOR_COMISSAO', 
		'TAXA_DESCONTO', 
		'VALOR_DESCONTO', 
		'VALOR_TOTAL', 
		'OBSERVACAO', 
	];
	
	static List<String> colunas = <String>[
		'Id', 
		'Código do Orçamento', 
		'Data de Cadastro', 
		'Data de Entrega', 
		'Data de Validade', 
		'Tipo do Frete', 
		'Valor Subtotal', 
		'Valor Frete', 
		'Taxa Comissão', 
		'Valor Comissão', 
		'Taxa Desconto', 
		'Valor Desconto', 
		'Valor Total', 
		'Observação', 
	];

	VendaOrcamentoCabecalho.fromJson(Map<String, dynamic> jsonDados) {
		id = jsonDados['id'];
		idVendaCondicoesPagamento = jsonDados['idVendaCondicoesPagamento'];
		idVendedor = jsonDados['idVendedor'];
		idCliente = jsonDados['idCliente'];
		idTransportadora = jsonDados['idTransportadora'];
		codigo = jsonDados['codigo'];
		dataCadastro = jsonDados['dataCadastro'] != null ? DateTime.tryParse(jsonDados['dataCadastro']) : null;
		dataEntrega = jsonDados['dataEntrega'] != null ? DateTime.tryParse(jsonDados['dataEntrega']) : null;
		validade = jsonDados['validade'] != null ? DateTime.tryParse(jsonDados['validade']) : null;
		tipoFrete = getTipoFrete(jsonDados['tipoFrete']);
		valorSubtotal = jsonDados['valorSubtotal'] != null ? jsonDados['valorSubtotal'].toDouble() : 0.0;
		valorFrete = jsonDados['valorFrete'] != null ? jsonDados['valorFrete'].toDouble() : 0.0;
		taxaComissao = jsonDados['taxaComissao'] != null ? jsonDados['taxaComissao'].toDouble() : 0.0;
		valorComissao = jsonDados['valorComissao'] != null ? jsonDados['valorComissao'].toDouble() : 0.0;
		taxaDesconto = jsonDados['taxaDesconto'] != null ? jsonDados['taxaDesconto'].toDouble() : 0.0;
		valorDesconto = jsonDados['valorDesconto'] != null ? jsonDados['valorDesconto'].toDouble() : 0.0;
		valorTotal = jsonDados['valorTotal'] != null ? jsonDados['valorTotal'].toDouble() : 0.0;
		observacao = jsonDados['observacao'];
		vendaCondicoesPagamento = jsonDados['vendaCondicoesPagamento'] == null ? null : VendaCondicoesPagamento.fromJson(jsonDados['vendaCondicoesPagamento']);
		vendedor = jsonDados['vendedor'] == null ? null : Vendedor.fromJson(jsonDados['vendedor']);
		cliente = jsonDados['cliente'] == null ? null : Cliente.fromJson(jsonDados['cliente']);
		transportadora = jsonDados['transportadora'] == null ? null : Transportadora.fromJson(jsonDados['transportadora']);
		listaVendaOrcamentoDetalhe = (jsonDados['listaVendaOrcamentoDetalhe'] as Iterable?)?.map((m) => VendaOrcamentoDetalhe.fromJson(m)).toList() ?? [];
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonDados = <String, dynamic>{};

		jsonDados['id'] = id ?? 0;
		jsonDados['idVendaCondicoesPagamento'] = idVendaCondicoesPagamento ?? 0;
		jsonDados['idVendedor'] = idVendedor ?? 0;
		jsonDados['idCliente'] = idCliente ?? 0;
		jsonDados['idTransportadora'] = idTransportadora ?? 0;
		jsonDados['codigo'] = codigo;
		jsonDados['dataCadastro'] = dataCadastro != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataCadastro!) : null;
		jsonDados['dataEntrega'] = dataEntrega != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataEntrega!) : null;
		jsonDados['validade'] = validade != null ? DateFormat('yyyy-MM-ddT00:00:00').format(validade!) : null;
		jsonDados['tipoFrete'] = setTipoFrete(tipoFrete);
		jsonDados['valorSubtotal'] = valorSubtotal;
		jsonDados['valorFrete'] = valorFrete;
		jsonDados['taxaComissao'] = taxaComissao;
		jsonDados['valorComissao'] = valorComissao;
		jsonDados['taxaDesconto'] = taxaDesconto;
		jsonDados['valorDesconto'] = valorDesconto;
		jsonDados['valorTotal'] = valorTotal;
		jsonDados['observacao'] = observacao;
		jsonDados['vendaCondicoesPagamento'] = vendaCondicoesPagamento == null ? null : vendaCondicoesPagamento!.toJson;
		jsonDados['vendedor'] = vendedor == null ? null : vendedor!.toJson;
		jsonDados['cliente'] = cliente == null ? null : cliente!.toJson;
		jsonDados['transportadora'] = transportadora == null ? null : transportadora!.toJson;
		

		var listaVendaOrcamentoDetalheLocal = [];
		for (VendaOrcamentoDetalhe objeto in listaVendaOrcamentoDetalhe ?? []) {
			listaVendaOrcamentoDetalheLocal.add(objeto.toJson);
		}
		jsonDados['listaVendaOrcamentoDetalhe'] = listaVendaOrcamentoDetalheLocal;
	
		return jsonDados;
	}
	
    getTipoFrete(String? tipoFrete) {
    	switch (tipoFrete) {
    		case 'C':
    			return 'CIF';
    		case 'F':
    			return 'FOB';
    		default:
    			return null;
    		}
    	}

    setTipoFrete(String? tipoFrete) {
    	switch (tipoFrete) {
    		case 'CIF':
    			return 'C';
    		case 'FOB':
    			return 'F';
    		default:
    			return null;
    		}
    	}


	String objetoEncodeJson(VendaOrcamentoCabecalho objeto) {
	  final jsonDados = objeto.toJson;
	  return json.encode(jsonDados);
	}
	
}